File: src/services/modbus.ts
import { invoke } from "@tauri-apps/api/core";
import { withLogging } from "@/services/logger";
import type {
  ModbusConnectionConfig,
  ModbusConnectionInfo,
  ModbusConnectionStatus,
  RegisterReadRequest,
  RegisterValue,
  RegisterWriteRequest,
  RegisterWriteResult,
  MonitorRequest,
  MonitoredRegister,
  ModbusPollResponse,
} from "@/types/modbus";

// ─── Connection ──────────────────────────────────────────────────

export const modbusConnect = withLogging(
  "modbus_connect",
  async (config: ModbusConnectionConfig): Promise<string> => {
    return invoke("modbus_connect", { config });
  }
);

export const modbusDisconnect = withLogging(
  "modbus_disconnect",
  async (connectionId: string): Promise<void> => {
    return invoke("modbus_disconnect", { connectionId });
  }
);

export const modbusGetConnections = withLogging(
  "modbus_get_connections",
  async (): Promise<ModbusConnectionInfo[]> => {
    return invoke("modbus_get_connections");
  }
);

export const modbusGetConnectionStatus = withLogging(
  "modbus_get_connection_status",
  async (connectionId: string): Promise<ModbusConnectionStatus> => {
    return invoke("modbus_get_connection_status", { connectionId });
  }
);

// ─── Register Read / Write ───────────────────────────────────────

export const modbusReadRegisters = withLogging(
  "modbus_read_registers",
  async (connectionId: string, request: RegisterReadRequest): Promise<RegisterValue[]> => {
    return invoke("modbus_read_registers", { connectionId, request });
  }
);

export const modbusWriteRegisters = withLogging(
  "modbus_write_registers",
  async (connectionId: string, request: RegisterWriteRequest): Promise<RegisterWriteResult[]> => {
    return invoke("modbus_write_registers", { connectionId, request });
  }
);

// ─── Monitor ─────────────────────────────────────────────────────

export const modbusAddMonitor = withLogging(
  "modbus_add_monitor",
  async (connectionId: string, request: MonitorRequest): Promise<MonitoredRegister> => {
    return invoke("modbus_add_monitor", { connectionId, request });
  }
);

export const modbusRemoveMonitor = withLogging(
  "modbus_remove_monitor",
  async (connectionId: string, monitorId: number): Promise<void> => {
    return invoke("modbus_remove_monitor", { connectionId, monitorId });
  }
);

export const modbusGetMonitors = withLogging(
  "modbus_get_monitors",
  async (connectionId: string): Promise<MonitoredRegister[]> => {
    return invoke("modbus_get_monitors", { connectionId });
  }
);

export const modbusPollMonitors = withLogging(
  "modbus_poll_monitors",
  async (connectionId: string): Promise<ModbusPollResponse> => {
    return invoke("modbus_poll_monitors", { connectionId });
  }
);

File: src-tauri/src/commands/modbus.rs
use crate::error::AppError;
use crate::modbus::types::*;
use crate::state::AppState;
use tauri::State;

// ─── Connection Commands ─────────────────────────────────────────

#[tauri::command]
pub async fn modbus_connect(
    state: State<'_, AppState>,
    config: ModbusConnectionConfig,
) -> Result<String, AppError> {
    state.modbus_manager.connect(config).await
}

#[tauri::command]
pub async fn modbus_disconnect(
    state: State<'_, AppState>,
    connection_id: String,
) -> Result<(), AppError> {
    state.modbus_manager.disconnect(&connection_id).await
}

#[tauri::command]
pub async fn modbus_get_connections(
    state: State<'_, AppState>,
) -> Result<Vec<ModbusConnectionInfo>, AppError> {
    Ok(state.modbus_manager.list_connections().await)
}

#[tauri::command]
pub async fn modbus_get_connection_status(
    state: State<'_, AppState>,
    connection_id: String,
) -> Result<ModbusConnectionStatus, AppError> {
    state.modbus_manager.get_status(&connection_id).await
}

// ─── Register Read / Write Commands ──────────────────────────────

#[tauri::command]
pub async fn modbus_read_registers(
    state: State<'_, AppState>,
    connection_id: String,
    request: RegisterReadRequest,
) -> Result<Vec<RegisterValue>, AppError> {
    state
        .modbus_manager
        .read_registers(&connection_id, &request)
        .await
}

#[tauri::command]
pub async fn modbus_write_registers(
    state: State<'_, AppState>,
    connection_id: String,
    request: RegisterWriteRequest,
) -> Result<Vec<RegisterWriteResult>, AppError> {
    state
        .modbus_manager
        .write_registers(&connection_id, &request)
        .await
}

// ─── Monitor Commands ────────────────────────────────────────────

#[tauri::command]
pub async fn modbus_add_monitor(
    state: State<'_, AppState>,
    connection_id: String,
    request: MonitorRequest,
) -> Result<MonitoredRegister, AppError> {
    state
        .modbus_manager
        .add_monitor(&connection_id, &request)
        .await
}

#[tauri::command]
pub async fn modbus_remove_monitor(
    state: State<'_, AppState>,
    connection_id: String,
    monitor_id: u32,
) -> Result<(), AppError> {
    state
        .modbus_manager
        .remove_monitor(&connection_id, monitor_id)
        .await
}

#[tauri::command]
pub async fn modbus_get_monitors(
    state: State<'_, AppState>,
    connection_id: String,
) -> Result<Vec<MonitoredRegister>, AppError> {
    state
        .modbus_manager
        .get_monitors(&connection_id)
        .await
}

#[tauri::command]
pub async fn modbus_poll_monitors(
    state: State<'_, AppState>,
    connection_id: String,
) -> Result<ModbusPollResponse, AppError> {
    state.modbus_manager.poll_monitors(&connection_id).await
}
